libzeth/circuits/mimc/mimc_round_gadget.tcc

```diff
@@ -0,0 +1,135 @@
+// Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+//
+// SPDX-License-Identifier: LGPL-3.0+
+
+#ifndef __ZETH_CIRCUITS_MIMC_ROUND_GADGET_TCC__
+#define __ZETH_CIRCUITS_MIMC_ROUND_GADGET_TCC__
+
+#include "libzeth/circuits/mimc/mimc_round_gadget.hpp"
+
+namespace libzeth
+{
+
+template<typename FieldT, size_t Exponent>
+MiMC_round_gadget<FieldT, Exponent>::MiMC_round_gadget(
+    libsnark::protoboard<FieldT> &pb,
+    const libsnark::pb_variable<FieldT> &x,
+    const libsnark::pb_variable<FieldT> &k,
+    const FieldT &c,
+    libsnark::pb_variable<FieldT> &result,
+    const bool add_k_to_result,
+    const std::string &annotation_prefix)
+    : libsnark::gadget<FieldT>(pb, annotation_prefix)
+    , _x(x)
+    , _k(k)
+    , _c(c)
+    , _result(result)
+    , _add_k_to_result(add_k_to_result)
+{
+    // Each condition requires an intermediate variable, except the final one,
+    // which uses _result (and optionally _k).
+    _exponents.resize(NUM_CONDITIONS - 1);
+}
+
+template<typename FieldT, size_t Exponent>
+void MiMC_round_gadget<FieldT, Exponent>::generate_r1cs_constraints()
+{
+    // Mask to capture the last bit
+    constexpr size_t mask = 1 << (EXPONENT_NUM_BITS - 1);
+    // t = x + k + c
+    libsnark::pb_linear_combination<FieldT> t;
+    t.assign(this->pb, _x + _k + _c);
+
+    // For first bit (1 by definition) compute t^2
+    size_t exp = Exponent << 1;
+    _exponents[0].allocate(this->pb, FMT(this->annotation_prefix, "^2"));
+    this->pb.add_r1cs_constraint(
+        libsnark::r1cs_constraint<FieldT>(t, t, _exponents[0]),
+        FMT(this->annotation_prefix, " mul[0]"));
```

Comment:
Happy to rename, but since this is a condition involving several variables I was reluctant to name it after a single variable.  These conditions are not stored as members, but It is supposed to be the "n-th multiplication".  How about something like `" calc_t^%zu"`?

Reply:
> How about something like " calc_t^%zu"?

Sounds good. Thanks.